File: js/apps/cms/repositories/structures.js
/*jslint indent: 2, nomen: true, maxlen: 100, white: true, plusplus: true, unparam: true */
/*global require, exports*/

////////////////////////////////////////////////////////////////////////////////
/// @brief A TODO-List Foxx-Application written for ArangoDB
///
/// @file This Document represents the repository communicating with ArangoDB
///
/// DISCLAIMER
///
/// Copyright 2010-2013 triagens GmbH, Cologne, Germany
///
/// Licensed under the Apache License, Version 2.0 (the "License");
/// you may not use this file except in compliance with the License.
/// You may obtain a copy of the License at
///
///     http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software
/// distributed under the License is distributed on an "AS IS" BASIS,
/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
/// See the License for the specific language governing permissions and
/// limitations under the License.
///
/// Copyright holder is triAGENS GmbH, Cologne, Germany
///
/// @author Michael Hackstein
/// @author Copyright 2011-2013, triAGENS GmbH, Cologne, Germany
////////////////////////////////////////////////////////////////////////////////

(function () {
  "use strict";
  
  var _ = require("underscore"),
    Foxx = require("org/arangodb/foxx"),
    structures = require("internal").db._collection("_structures"); // TODO!
  exports.Repository = Foxx.Repository.extend({
    // Define the save functionality
    getMonitored: function () {
      return [{
          name: "String"
        },{
          name: "Number"
        }];
      /*
      var a = structures.all(),
        res = [];
      while (a.hasNext()) {
        res.push({
          name: a.next().collection
        });
      }
      return res;
      */
    },
    
    getStructs: function (name) {
      switch(name) {
        case "String":
          return {
            attributes: {
              name : {
                type: "string"
              }
            }
          };
          break;
        case "Number":
          return {
            attributes: {
              int : {
                type: "number"
              },
              double: {
                type: "number"
              },
              boolean: {
                type: "boolean"
              }
            }
          };
          break;
        case "All":
          return {
            attributes: {
              int : {
                type: "number"
              },
              double: {
                type: "number"
              },
              name : {
                type: "string"
              },
              boolean: {
                type: "boolean"
              }
            }
          };
          break;
        default:
          res.json("Error has to be done!");
      }
      /*
      return structures.firstExample({collection: name});
      */
    },
    
    getContent: function (name) {
      switch(name) {
        case "String":
          return [{
            _id: "String/1",
            _rev: "1",
            _key: "1",
            name: "Test"
          }];
        case "Number":
          return [{
            _id: "Number/1",
            _rev: "1",
            _key: "1",
            int: 4,
            double: 4.5,
            boolean: true
          }];
        case "All":
          return [{
            _id: "All/1",
            _rev: "1",
            _key: "1",
            int: 4,
            double: 4.5,
            name: "Test",
            boolean: true
          }];
        default:
          return "Error has to be done!";
      }
      /*
      var db = require("internal").db._collection(name);
      // TODO: Requires Optimization!
      return db.toArray();
      */
    }
  });
  
}());

File: lib/Basics/structures.cpp
////////////////////////////////////////////////////////////////////////////////
/// DISCLAIMER
///
/// Copyright 2014-2016 ArangoDB GmbH, Cologne, Germany
/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
///
/// Licensed under the Apache License, Version 2.0 (the "License");
/// you may not use this file except in compliance with the License.
/// You may obtain a copy of the License at
///
///     http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software
/// distributed under the License is distributed on an "AS IS" BASIS,
/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
/// See the License for the specific language governing permissions and
/// limitations under the License.
///
/// Copyright holder is ArangoDB GmbH, Cologne, Germany
///
/// @author Dr. Frank Celler
////////////////////////////////////////////////////////////////////////////////

#include "Basics/Common.h"

////////////////////////////////////////////////////////////////////////////////
/// @brief destroys the data of blob, but does not free the pointer
////////////////////////////////////////////////////////////////////////////////

void TRI_DestroyBlob(TRI_memory_zone_t* zone, TRI_blob_t* blob) {
  if (blob != nullptr) {
    if (blob->data != nullptr) {
      TRI_Free(zone, blob->data);
    }
  }
}

////////////////////////////////////////////////////////////////////////////////
/// @brief copies a blob into given destination
////////////////////////////////////////////////////////////////////////////////

int TRI_CopyToBlob(TRI_memory_zone_t* zone, TRI_blob_t* dst,
                   TRI_blob_t const* src) {
  dst->length = src->length;

  if (src->length == 0 || src->data == nullptr) {
    dst->length = 0;
    dst->data = nullptr;
  } else {
    dst->data = static_cast<char*>(TRI_Allocate(zone, dst->length, false));

    if (dst->data == nullptr) {
      return TRI_ERROR_OUT_OF_MEMORY;
    }

    memcpy(dst->data, src->data, src->length);
  }

  return TRI_ERROR_NO_ERROR;
}

////////////////////////////////////////////////////////////////////////////////
/// @brief assigns a blob value by reference into given destination
////////////////////////////////////////////////////////////////////////////////

int TRI_AssignToBlob(TRI_memory_zone_t* zone, TRI_blob_t* dst,
                     TRI_blob_t const* src) {
  dst->length = src->length;

  if (src->length == 0 || src->data == nullptr) {
    dst->length = 0;
    dst->data = nullptr;
  } else {
    dst->length = src->length;
    dst->data = src->data;
  }

  return TRI_ERROR_NO_ERROR;
}
